### PR TITLE
Customizable impact victim scan radius

### LINKED ISSUE #9173: Something is off about warhead spread/health radius interaction
zero-damage

Health radius `1c0` on the cyard, `Spread: 96` on the `QuadRockets`, gave `BlocksProjectiles` trait to cyard, missiles explodes barely within health radius, zero damage. 
As far as I understand the concept of the health radius and spread, it should deal full damage within 96 wdist units (3 pixel in d2k) of the outer border of the health radius, 37% at 192 wdist/6px etc. 

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/26212526-something-is-off-about-warhead-spread-health-radius-interaction?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by reaperrr**
Fixes #9173.

### CODE REVIEW

File: OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs [hunk 1]
```diff
@@ -19,6 +19,9 @@ public class SpreadDamageWarhead : DamageWarhead
 		[Desc("Range between falloff steps.")]
 		public readonly WDist Spread = new WDist(43);
 
+		[Desc("Extra search radius beyond maximum spread. Required to ensure damage to actors with large health radius.")]
+		public readonly WDist VictimExtraSearchRadius = new WDist(2048);
```
Comment (pchote) [inline OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs]: Can we get a lint rule that checks all of these vs all of the health radii?

Comment (reaperrr) [inline OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs]: Makes sense, I guess. I'll give it a try.

**Comment by Unit158** [comment]
Why would the raycaster help? I'll be happy to get that code up and running again for you. I'll just need a little bit of your brain in exchange :p

**Comment by reaperrr** [comment]
Rebased and added lint rule.

Note: The Travis error is caused by the lint testcase to demonstrate that the rule works.

File: OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs (hunk 1 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs]: Done.

**Comment by reaperrr** [comment]
Updated, renamed `Arm` to `ArmDelay` in `Bullet` and fixed its height check in `shouldExplode`.

**Comment by reaperrr** [comment]
I've removed the lint testcase since the failed travis annoyed me, but you can easily test it:

- give any `SpreadDamageWarhead` a custom `VictimExtraSearchRadius` with a value of less than `426`, but more than `128` (which is the `Health.Radius` default), then `make test`. It should fail on any non-infantry actor.
- give any actor a `Health.Radius` of more than `2c0`, then the test will fail on every `SpreadDamageWarhead` on that actor since the default `VictimExtraSearchRadius` is less than that.

File: OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs (hunk 1 above)
Comment (pchote) [inline OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs]: Can we call this `Target` instead of `Victim` for consistency?

**Comment by pchote** [comment]
Looks sensible overall, but I haven't tested it.  Needs a rebase.

**Comment by reaperrr** [comment]
Rebased.

**Comment by reaperrr** [comment]
I'm beginning to think I should go for a `ExplodeOnContactType` instead, with (initial) available types being 
- `ContactType.Blocker` (actors with BlocksProjectiles, would replace `Blockable` boolean), 
- `ContactType.EnemyHealthRadius` (if you want this to trigger only on enemies), 
- `ContactType.AlliedHealthRadius` (if you want this to only trigger on allies), 
- `ContactType.HealthRadius` (if you want this to trigger on any health radius). 

We could then later add more types like `.Shield` (makes the projectile blow up outside the health radius and only damage the shield).


If I do that, it would probably be better to split that off to its own PR and only deal with the warhead target check radius in this PR.

Thoughts?

**Comment by UberWaffe** [comment]
@reaperrr Would you need ray-casting between two WPos to do some of these?
There is a working & tested (but only by myself) ray-casting function in the pending pull request  #9357.

It currently only takes the actor's health radius into account, along with the 'width' of the beam, but it has proven to work well (touch wood) for making the Line Impact Projectiles stop / impact blocking actors regardless of the speed of the Line Impact Projectile.
See disruptor beam (slow moving) and ghost stalker railgun (instant hit) for test cases.
I'm not entirely happy with how I 'redirect' them when being blocked, but that is not a fault of the ray-caster itself, but rather the projectile code.

**Comment by atlimit8** [comment]
Needs to be rebased.

**Comment by reaperrr** [comment]
Rebased.

Additionally, I decided to split off the ExplodeOnContact logic (will file a PR later) since that might take longer, and this here is both a simpler feature and a bugfix as well, so it should not be delayed any longer than necessary.

File: OpenRA.Mods.Common/Effects/Bullet.cs [hunk 2]
```diff
@@ -164,8 +164,14 @@ public void Tick(World world)
 			if (info.ContrailLength > 0)
 				contrail.Update(pos);
 
-			if (ticks++ >= length || (info.Blockable && world.ActorMap
-				.GetUnitsAt(world.Map.CellContaining(pos)).Any(a => a.Info.HasTraitInfo<IBlocksProjectilesInfo>())))
+			var cell = world.Map.CellContaining(pos);
+			var height = world.Map.DistanceAboveTerrain(pos);
+
+			var shouldExplode = (height.Length <= 0) // Hit the ground
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Effects/Bullet.cs]: Unnecessary brackets.

**Comment by reaperrr** [comment]
Updated.

**Comment by penev92** [comment]
Looks reasonable :+1:

**Comment by RoosterDragon** [comment]
:+1: Code looks good and confirmed this fixes your reproduction case from #9173..

**Comment by RoosterDragon** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/4b93e7fdf229588fd7c18a1f45de7485db7223fc...8e31fb4bbbfbb7ef7a202a821fca4d3dc02f25e6)

**Comment by reaperrr** [comment]
The Bullet's new `ArmDelay` breaks several instant-hit weapons (just watch RA shellmap), I'll file a fix shortly.

**Comment by Mailaender** [comment]
This check breaks for https://github.com/OpenRA/OpenRA/pull/10613. I need help as I don't understand it.

**Comment by GraionDilach** [comment]
`TargetExtraSearchRadius` has been decreased to 1536 since then, but since the gates have a diagonal of sqrt2*1.5 cells (2173 wu), all warhead TESR which can affect it needs to be bigger than the aforementioned 2173.

**Comment by reaperrr** [comment]
> EDIT: Oh wait, Radius, these needs to be doubled to 4346 as the minimum, -_-

Nope, `TargetExtraSearchRadius` is compared against `HitShape.OuterRadius` so it's radius vs. radius, no need to double them.

Also, `2048` is enough according to my tests.

**Pull request merged**